From 2c72a51cd47ad7d4bf9ef8752be93094c8deb58c Mon Sep 17 00:00:00 2001
From: Manikanta Mylavarapu <quic_mmanikan@quicinc.com>
Date: Thu, 28 Sep 2023 12:32:40 +0530
Subject: [PATCH 416/500] dmaengine: qcom: bam_dma: update mask field value for
 read

Currently mask field is configured with 0xFFFFFFFF
for both write & read. But it's not required
for read operation. In 32-bit address, mask field
is reserved and in 36-bit address scheme, 4 LSB
bits of mask field only used. Rest of [31:4]bits
are reserved.

Devsoc supports 36-bit address scheme. So if we
configure mask with 0xFFFFFFFF, QPIC controller
adds 4 bits of mask field value (0xF) to address.
It leads to wrong DDR address.

This patch resolves issue by configuring mask
field with 0x0 for read operation.

Command element structure

1st	-----------------------------------------
DWORD   |Command opcode |Peripheral address	|
	-----------------------------------------
	31	      24 23			0

2nd	-----------------------------------------
DWORD   | Write: Data to be written		|
	| Read: 32 LSB bits of address		|
	-----------------------------------------
	31					0

3rd	--------------------------------------------
DWORD	| Write: Mask value (0xFFFFFFFF)|4 LSB bits|
(Mask)	| Read: Reserved		|of address|
	--------------------------------------------
	31				4 3	   0

4th	-----------------------------------------
DWORD	|	Reserved			|
	-----------------------------------------
	31					0

Change-Id: I7b42450abf60724caa8609f1c6d95d8ecf881a4a
Signed-off-by: Manikanta Mylavarapu <quic_mmanikan@quicinc.com>
---
 include/linux/dma/qcom_bam_dma.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/dma/qcom_bam_dma.h b/include/linux/dma/qcom_bam_dma.h
index 68fc0e643b1b..0d55c019cd8e 100644
--- a/include/linux/dma/qcom_bam_dma.h
+++ b/include/linux/dma/qcom_bam_dma.h
@@ -67,5 +67,7 @@ bam_prep_ce(struct bam_cmd_element *bam_ce, u32 addr,
 	    enum bam_command_type cmd, u32 data)
 {
 	bam_prep_ce_le32(bam_ce, addr, cmd, cpu_to_le32(data));
+	/* For read operation, mask field is reserved */
+	bam_ce->mask = 0x0;
 }
 #endif
-- 
2.34.1

